### ISSUE TITLE
Authentication Failure in Copilot Mode: Missing Headers

### ISSUE THREAD

**Issue by lianghsun** [Chainlit/chainlit#771]
I've been considering that anything related to tokens would likely utilize `header_auth_callback(headers: Dict)`. I attempted to create a mock login (which guarantees authentication success regardless of the situation) to capture headers sent by the nginx reverse proxy. This approach works in app mode, but fails to receive headers in copilot mode, as illustrated in the attached screenshot:

**mock login**
![截圖 2024-02-27 上午10 50 02](https://github.com/Chainlit/chainlit/assets/25081193/da14bd99-f617-4fcb-bb22-f52acce2d3cd)

**App mode do return headers**
![截圖 2024-02-27 上午11 06 08](https://github.com/Chainlit/chainlit/assets/25081193/fe6c6a77-d989-4343-9eb7-d2444a48cd26)

Using the developer tools, I noticed an error message at `index.js:78`, indicating an unauthorized access attempt:
```
GET http://localhost:8889/project/settings?language=zh-TW 401 (Unauthorized)
```
**Copilot mode**
<img width="1040" alt="截圖 2024-02-27 上午10 49 19" src="https://github.com/Chainlit/chainlit/assets/25081193/b49f14f3-f651-4e00-974e-25ad9fced622">

**App mode**
<img width="1040" alt="截圖 2024-02-27 上午10 49 01" src="https://github.com/Chainlit/chainlit/assets/25081193/e0d0f441-fac0-4d23-8025-b89ec89a56df">

This issue does not occur in app mode, only in copilot mode. The current problem is that I intend to use headers for authentication in copilot, but the server side does not receive any headers, thus failing to verify identity. Even with mock login (which should ensure authentication success), my simple HTML page displays an unauthorized access error.

Could you advise on how to resolve this issue? Thank you.

**Comment by karankapur999** [comment]
Hi 
I am facing the same error!!
Did you got it working?

**Comment by lianghsun** [comment]
@karankapur999 Not yet

**Comment by ajosegun** [comment]
Hi,

I am facing a similar error!!

Did you get it working?

**Comment by HarshJa1n** [comment]
facing the same error, anybody got it working?

**Comment by chenm1xuexi** [comment]
That's so bad! I'm facing the same error. I found that Copilot uses WebSocket, while the web uses HTTP. Perhaps the error occurs because of this

**Comment by stephenrs** [comment]
Authentication works differently in copilot mode than when using the full app. Instead of relying on the header_auth_callback, you have to supply an access_token via your frontend Javascript in the call to window.mountChainlitWidget(). header_auth_callback is not called in copilot mode.

You can see access_token in the widget configuration section here: https://docs.chainlit.io/deploy/copilot

You can see how to create the access_token in the Authentication section at the bottom of that page.

**Comment by dokterbob** [comment]
> Authentication works differently in copilot mode than when using the full app. Instead of relying on the header_auth_callback, you have to supply an access_token via your frontend Javascript in the call to window.mountChainlitWidget(). header_auth_callback is not called in copilot mode.
> 
> You can see access_token in the widget configuration section here: https://docs.chainlit.io/deploy/copilot
> 
> You can see how to create the access_token in the Authentication section at the bottom of that page.
> 
> This is also discussed here: #756

We just discussed that we really kind of want to move away from this custom approach towards supporting:
1. Header-based auth (e.g. behind a proxy).
2. OAuth (e.g. 3rd party auth is done through normal OAuth).

I'll keep you posted.

**Comment by alex-bgamestudios** [comment]
any updates here?? copilot auth kinda sucks

**Comment by shutttttdown** [comment]
> Hi [@lianghsun](https://github.com/lianghsun), I am not able to get it working in the web app as well, how did you do it?

Same error even for the newest chainlit verion(2.5.0),

any update?

Than you.

**Comment by holmbergjonas** [comment]
So issues are closed if not tended to for 21 days? Great way to close tickets, just ignore them....

**Comment by stephenrs** [comment]
Yes, this is exactly the kind of thing that made me lose interest in this project. The ethics and competency of the maintainers are highly questionable. It's hard to trust what's going on here. 

As you can see on this issue, they have historically used bots to mask/hide long-standing problems that they have simply refused to address. This has the intended effect of keeping the number in the issues tab artificially lower than the true number of outstanding issues. Be careful with this project.
